File: Contents/InitialConsonantSearch/IndicatorSettingsTest/Main/TEST/GeometricShapesView.m
//
//  GeometricShapesView.m
//  ToolSettingTest
//
//  Created by Kwan Hyun Son on 2023/09/15.
//

#import "GeometricShapesView.h"

@interface GeometricShapesView ()
@property (nonatomic, strong, readonly) CAShapeLayer *shapeLayer; // @dynamic
@end

@implementation GeometricShapesView
@dynamic shapeLayer;

+ (Class)layerClass {
    return [CAShapeLayer class];
}

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        CommonInit(self);
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)coder {
    self = [super initWithCoder:coder];
    if (self) {
        
    }
    return self;
}

- (void)awakeFromNib {
    [super awakeFromNib];
    CommonInit(self);
}

- (void)layoutSubviews {
    [super layoutSubviews];
    [self drawGeometricShapes];
}

#pragma mark - 생성 & 소멸

- (instancetype)initWithType:(GeometricShapesViewType)type {
    self = [self initWithFrame:CGRectZero];
    if (self) {
        _geometricShapesType = type;
    }
    return self;
}

static void CommonInit(GeometricShapesView *self) {
    self->_borderColor = [UIColor blackColor];
    self->_borderWidth = 2.0;
    self->_backColor = [UIColor cyanColor];
}

#pragma mark - 세터 & 게터
- (CAShapeLayer *)shapeLayer {
    return (CAShapeLayer *)(self.layer);
}

- (void)setBorderColor:(UIColor *)borderColor {
    _borderColor = borderColor;
    [self setNeedsLayout];
}

- (void)setBorderWidth:(CGFloat)borderWidth {
    _borderWidth = borderWidth;
    [self setNeedsLayout];
}

- (void)setBackColor:(UIColor *)backColor {
    _backColor = backColor;
    [self setNeedsLayout];
}

- (void)setGeometricShapesType:(GeometricShapesViewType)geometricShapesType {
    _geometricShapesType = geometricShapesType;
    [self setNeedsLayout];
}

#pragma mark - Actions

- (void)drawGeometricShapes {
    if (self.geometricShapesType == GeometricShapesViewTypeEllipse) {
        [self drawEllipse];
    } else if (self.geometricShapesType == GeometricShapesViewTypeRectangle) {
        [self drawRectangle];
    } else if (self.geometricShapesType == GeometricShapesViewTypeTriangle) {
        [self drawTriangle];
    } else {
        NSCAssert(FALSE, @"정상이 아닌 값이 들어왔다");
    }
    
    CAShapeLayer *shapeLayer = self.shapeLayer;
    shapeLayer.strokeColor = self.borderColor.CGColor;
    shapeLayer.lineWidth = self.borderWidth;
    shapeLayer.fillColor = self.backColor.CGColor;
}

- (void)drawEllipse {
    UIBezierPath *path = [UIBezierPath bezierPathWithOvalInRect:self.bounds];
    self.shapeLayer.path = path.CGPath;
}

- (void)drawRectangle {
    UIBezierPath *path = [UIBezierPath bezierPathWithRect:self.bounds];
    self.shapeLayer.path = path.CGPath;
}

- (void)drawTriangle {
    CGPoint p1 = CGPointMake(self.bounds.size.width / 2.0, 0.0);
    CGPoint p2 = CGPointMake(0.0, self.bounds.size.height);
    CGPoint p3 = CGPointMake(self.bounds.size.width, self.bounds.size.height);
    UIBezierPath *path = [UIBezierPath bezierPath];
    [path moveToPoint:p1];
    [path addLineToPoint:p2];
    [path addLineToPoint:p3];
    [path closePath];
    self.shapeLayer.path = path.CGPath;
}

@end

File: Contents/InitialConsonantSearch/InitialConsonantSearch_Swift/Main/TEST/GeometricShapesView.swift
//
//  GeometricShapesView.swift
//  ToolSettingTest_Swift
//
//  Created by Kwan Hyun Son on 2023/09/15.
//

import UIKit

extension GeometricShapesView {
    
    // MARK: - Nested Type
    
    enum Kind: Int {
        case ellipse = 0 // 타원
        case rectangle   // 사각형
        case triangle    // 삼각형
    }
}

final class GeometricShapesView: UIView {
    
    // MARK: - Property
    
    var geometricShapesType: Kind = .ellipse {
        didSet {
            setNeedsLayout()
        }
    }
    var borderColor: UIColor = .black {
        didSet {
            setNeedsLayout()
        }
    }
    var borderWidth: CGFloat = 2.0 {
        didSet {
            setNeedsLayout()
        }
    }
    var backColor: UIColor = .cyan {
        didSet {
            setNeedsLayout()
        }
    }
    
    private var shapeLayer: CAShapeLayer {
        return layer as! CAShapeLayer
    }
    
    // MARK: - Override
    
    override class var layerClass: AnyClass {
        return CAShapeLayer.self
    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        commonInit()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    override func awakeFromNib() {
        super.awakeFromNib()
        commonInit()
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        drawGeometricShapes()
    }
    
    // MARK: - 생성 & 소멸
    
    convenience init(type: Kind) {
        self.init(frame: CGRect.zero)
        geometricShapesType = type
    }
    
    private func commonInit() { }
    
    // MARK: - Actions
    
    private func drawGeometricShapes() {
        switch geometricShapesType {
        case .ellipse:
            drawEllipse()
        case .rectangle:
            drawRectangle()
        case .triangle:
            drawTriangle()
        }
        
        shapeLayer.strokeColor = borderColor.cgColor
        shapeLayer.lineWidth = borderWidth
        shapeLayer.fillColor = backColor.cgColor
    }
    
    private func drawEllipse() {
        let path = UIBezierPath(ovalIn: bounds)
        shapeLayer.path = path.cgPath
    }
    
    private func drawRectangle() {
        let path = UIBezierPath(rect: bounds)
        shapeLayer.path = path.cgPath
    }
    
    private func drawTriangle() {
        let p1 = CGPoint(x: bounds.width / 2.0, y: 0.0)
        let p2 = CGPoint(x: 0.0, y: bounds.height)
        let p3 = CGPoint(x: bounds.width, y: bounds.height)
        let path = UIBezierPath()
        path.move(to: p1)
        path.addLine(to: p2)
        path.addLine(to: p3)
        path.close()
        shapeLayer.path = path.cgPath
    }
}

